### PR TITLE
Add integration and docs for Django Ninja authentication

### LINKED ISSUE #1373: Add a documentation how to integrate `Django-ninja`
**Is your feature request related to a problem? Please describe.**
We're currently using django-oauth-toolkit together with DRF and we want to consider a switch to https://django-ninja.dev.

**Describe the solution you'd like**
Either some helper/utils function/classes to easily integrate Django-ninja and/or a documentation page (similar to https://django-oauth-toolkit.readthedocs.io/en/latest/rest-framework/rest-framework.html)

**Describe alternatives you've considered**
Haven't considered any alternatives yet.

### PR DESCRIPTION
**Pull request by brianhelba**
Fixes #1373.

## Checklist
- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] `CHANGELOG.md` updated (only for user relevant changes)
- [x] author name in `AUTHORS`
- [ ] tests/app/idp updated to demonstrate new features
- [ ] tests/app/rp updated to demonstrate new features

### CODE REVIEW

File: docs/ninja.rst [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        ...
+
+
+Custom Authorization Behavior
+-----------------------------
+``HttpOAuth2`` can be extended to provide custom authorization behaviors.
+
+Simply subclass it and override its ``authenticate`` method.
+
+.. autoclass:: oauth2_provider.contrib.ninja.HttpOAuth2
```
Comment (brianhelba) [inline docs/ninja.rst:101]: I'd like to test how this renders, but the docs build process is broken for me locally.

Comment (dopry) [reply docs/ninja.rst:101]: You can view the build output if you're having trouble locally.

File: oauth2_provider/contrib/ninja/security.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def __init__(self, *, scopes: list[str] | None = None) -> None:
+        super().__init__()
+        self.scopes = scopes if scopes is not None else []
+
+    def __call__(self, request: HttpRequest) -> Any | None:
+        oauthlib_core = get_oauthlib_core()
+        # This also sets `request.user`,
+        # which Ninja does not: https://github.com/vitalik/django-ninja/issues/76
+        valid, r = oauthlib_core.verify_request(request, scopes=self.scopes)
```
Comment (brianhelba) [inline oauth2_provider/contrib/ninja/security.py:23]: I originally copied this logic: https://github.com/django-oauth/django-oauth-toolkit/blob/dbedfaa742ec7d38caa0f7be73b571782fb13505/oauth2_provider/contrib/rest_framework/authentication.py#L30-L35

However, I don't actually think that's necessary here, since this only supports obtaining the access token via an HTTP header and not via a query string.

If a user does something illegal enough to make `oauthlib_core.verify_request` raise an error, there's no need to convert it into a `SuspiciousOperation` if it's specifically a hex encoding issue (since URL encoding of HTTP headers isn't a common client mistake).

File: docs/ninja.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+
+Custom Authorization Behavior
+-----------------------------
+``HttpOAuth2`` can be extended to provide custom authorization behaviors.
+
+Simply subclass it and override its ``authenticate`` method.
+
+.. autoclass:: oauth2_provider.contrib.ninja::HttpOAuth2
+   :members: authenticate
```
Comment (brianhelba) [inline docs/ninja.rst]: I'm not able to get Sphinx to build this. It's [producing errors](https://app.readthedocs.org/api/v2/build/31248052.txt) like:
> WARNING: autodoc: failed to import class 'ninja.HttpOAuth2' from module 'oauth2_provider.contrib'; the following exception was raised:
No module named 'ninja'

I've tried changing it to `oauth2_provider.contrib.ninja.security::HttpOAuth2`, `oauth2_provider.contrib.ninja.security.HttpOAuth2`, etc. with no success.

I'd propose just keeping this here. Currently, it just doesn't render at all (so there's no visible breakage for users), but I'm hoping that an upgrade of Sphinx will magically fix it in the future.

File: oauth2_provider/contrib/ninja/security.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py:24]: False positive. See comment: https://github.com/django-oauth/django-oauth-toolkit/pull/1646#discussion_r2757014404

File: oauth2_provider/contrib/ninja/security.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        if not valid:
+            return None
+
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py:27]: False positive. See comment: https://github.com/django-oauth/django-oauth-toolkit/pull/1646#discussion_r2757014404

File: docs/ninja.rst (hunk 3 above)
Comment (dopry) [reply docs/ninja.rst]: Looking at the repo it looks like maybe it should be `oauth2_provider.contrib.ninja.HttpOAuth2`

File: oauth2_provider/contrib/ninja/security.py (hunk 2 above)
Comment (dopry) [reply oauth2_provider/contrib/ninja/security.py:23]: I'm not familiar with Ninja, are there any ninja docs about how to handle errors in this context?

File: CHANGELOG.md [hunk 6]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 * #1637 Support for Django 6.0
+* #1646 Integration and docs for Django Ninja authentication
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply CHANGELOG.md]: @dopry What do you want the changelog to reference: the issue or the PR?

File: tox.ini [hunk 7]
```diff
@@ -47,6 +47,7 @@ deps =
     dj60: Django>=6.0,<6.1
     djmain: https://github.com/django/django/archive/main.tar.gz
     djangorestframework
+    django-ninja
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply tox.ini:50]: This is just a test dependency, so if `django-ninja` releases something incompatible, it's actually helpful that the tests will automatically pick it up and alert us to new failures.

It'd also be reasonable if test determinism was prioritized, but none of the other test dependencies do this, and adding a lockfile is well out of scope of this PR.

File: oauth2_provider/contrib/ninja/security.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+# Don't inherit from `HttpBearer`, since we have our own header extraction logic
+class HttpOAuth2(HttpAuthBase):
+    """Perform OAuth2 authentication, for use with Django Ninja."""
+
+    openapi_scheme: str = "bearer"
+
+    def __init__(self, *, scopes: list[str] | None = None) -> None:
+        super().__init__()
+        self.scopes = scopes if scopes is not None else []
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py]: Reasonable point, I'll change this.

File: docs/ninja.rst [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        ...
+
+
+Custom Authorization Behavior
+-----------------------------
+``HttpOAuth2`` can be extended to provide custom authorization behaviors.
+
+Simply subclass it and override its ``authenticate`` method.
+
+.. autoclass:: oauth2_provider.contrib.ninja::HttpOAuth2
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply docs/ninja.rst]: I think I tried it, but I'll try again.

File: docs/ninja.rst (hunk 3 above)
Comment (brianhelba) [reply docs/ninja.rst]: I think I tried it, but I'll try again.

File: oauth2_provider/contrib/ninja/security.py (hunk 2 above)
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py:23]: [This section](https://django-ninja.dev/guides/authentication/#custom-exceptions) is the only docs.

Looking at the code, if an auth handler raises an exception, [it'll be caught](https://github.com/vitalik/django-ninja/blob/56e868754e4ede8a47beb1833c8f67d4134339ea/ninja/operation.py#L554-L555) and the overall auth flow will immediately fail (and not proceed to any subsequent checks).

I think allowing any exception from `oauthlib_core.verify_request` to propagate is reasonable, as it indicates not just a simple failure, but something structurally wrong or suspicious, so we should probably fail the overall access attempt.

The only alternative would be to catch all or certain exceptions coming from `oauthlib_core.verify_request` and `return None` instead, indicating a graceful failure and giving any subsequent auth handlers a chance to run.

File: tests/test_ninja.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+Application = get_application_model()
+AccessToken = get_access_token_model()
+UserModel = get_user_model()
+
+api = NinjaAPI()
+
+
+@api.get("/private", auth=HttpOAuth2())
+def private_endpoint(request):
+    return {"message": "This is a private endpoint"}
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply tests/test_ninja.py:25]: Good idea, I'll do it.

[the remark this answers is not in the document]
Comment (brianhelba) [reply tests/test_ninja.py:25]: This actually caught a bug in my earlier assumptions about `oauthlib_core`. Fixed the bug and added the test.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dopry) [reply CHANGELOG.md]: If the PR completely closes the issue, then the issue. If is closes some subset of the issue, then the PR and mention that it partially completes the issue.

File: oauth2_provider/contrib/ninja/security.py (hunk 2 above)
Comment (dopry) [reply oauth2_provider/contrib/ninja/security.py:23]: was this something copilot brought up, or just something you commented on reviewing yourself?

Comment (dopry) [reply oauth2_provider/contrib/ninja/security.py:23]: This looks fine.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (brianhelba) [reply CHANGELOG.md]: This totally resolves the issue, IMO. I'll change the reference to the issue #1373 then.

File: docs/ninja.rst (hunk 3 above)
Comment (dopry) [reply docs/ninja.rst]: I think ninja was missing in the deps during the doc build.

File: tests/test_ninja.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    def test_valid_token_with_scope(self):
+        auth = self._create_authorization_header(self.access_token.token)
+        response = self.client.get("/api/write", HTTP_AUTHORIZATION=auth)
+        self.assertEqual(response.status_code, 200)
+
+    def test_valid_token_absent_scope(self):
+        auth = self._create_authorization_header(self.access_token.token)
+        response = self.client.get("/api/impossible", HTTP_AUTHORIZATION=auth)
+        self.assertEqual(response.status_code, 401)
```
[the remark this answers is not in the document]
Comment (dopry) [reply tests/test_ninja.py:102]: the code path doesn't exist in the ninja implementation.

File: oauth2_provider/contrib/ninja/security.py (hunk 2 above)
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py:23]: Both? I originally copied the logic somewhat blindly, then changed my mind when I looked more closely at how it was being used. I thought it warranted a note in case reviewers wondered the same.

Copilot [brought this up independently](https://github.com/django-oauth/django-oauth-toolkit/pull/1646#discussion_r2757093833), as apparently it doesn't yet read my comments. :cry:

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (brianhelba) [reply CHANGELOG.md]: Just pushed another commit fixing the `CHANGELOG`.

File: docs/ninja.rst (hunk 3 above)
Comment (brianhelba) [reply docs/ninja.rst]: Ah, that might be it, thanks!

I'm fine with the reference as-is, but I actually think `oauth2_provider.contrib.ninja::HttpOAuth2` is the most correct syntax, as it disambiguates that `HttpOAuth2` is a class inside the `ninja` module, not another module itself.

See [the Tip here](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#usage).

**Review by dopry — APPROVED** [review]

**Comment by brianhelba** [comment]
@dopry Thanks for the help reviewing this!

If any further feedback comes up, I'm happy to address it on March 16th (I'll be offline until then).

**Review by dopry — CHANGES_REQUESTED** [review]
@brianhelba  I'm concerned about the _call issue copilot raised. Can you look into that?

File: oauth2_provider/contrib/ninja/security.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+        if not valid:
+            return None
+
+        # Ninja doesn't automatically set `request.user`: https://github.com/vitalik/django-ninja/issues/76
+        # However, Django's AuthenticationMiddleware (which does set this from a session cookie) is
+        # ubiquitous, and even Ninja's own tutorials assume that `request.user` will somehow be set,
+        # so ensure that authentication via OAuth2 doesn't violate expectations.
+        request.user = r.user
+
```
[the remark this answers is not in the document]
Comment (brianhelba) [reply oauth2_provider/contrib/ninja/security.py:33]: This is an interesting consideration, but Copilot's interpretation is incorrect.

To restate my code comment, as Copilot doesn't seem to understand... At this point in the code `valid is True`. We should assume that our `HttpOAuth2` handler is responsible for authenticating the entirety of this request (unfortunately, there's no specification to follow, but this seems like the only behavior that makes sense). It's possible that a session cookie was also passed with this request, in addition to OAuth2 `Authorization: Bearer ...` headers; in that case, an `AuthenticationMiddleware` would have pre-set `request.user` to whatever the session cookie resolves, but we want to always overwrite that to whatever the OAuth2 token resolves. We have to do this unconditionally, or else we'd have conflicting sources of truth for the request user.

However, `r` is an `AbstractAccessToken`, and [`r.user` is nullable](https://github.com/django-oauth/django-oauth-toolkit/blob/a7e114eff7600573d3e6cc4ab7acf1c8451e4074/oauth2_provider/models.py#L394
). What should we do if `r.user is None`? The answer should not depend on whether an `AuthenticationMiddleware` has already set `request.user`, because we now "own" the request's authentication and should unconditionally set the user.

I don't think we should set `request.user = AnonymousUser()`, because this isn't an anonymous request. **It's an authenticated request**, but the token simply doesn't have an associated user (which OAuth-Toolkit apparently allows).

However, while `Request.user` isn't a part of pure Django (it's not defined on the `Request` model), practically it's always patched on by `AuthenticationMiddleware` (which is typically enabled). So, most sensible projects ([and `django-stubs`](https://github.com/typeddjango/django-stubs/blob/2f83b71086905db731a78afacb81d5c6b7ee81aa/django-stubs/http/request.pyi#L62-L63)) assume that `request.user` is always set to something: either an actual user or `AnonymousUser` (it is **not** expected to be `None`, so people don't guard against calling something like `request.user.is_authenticated`).

So, what do we do? If you force me to resolve the compromise, I'd say perhaps we should set it to `AnonymousUser()`, so at least we won't break people's type expectations, plus I'm hoping that having `AbstractAccessToken.user` be `None` is pretty rare. However, I'd really appreciate some additional feedback @dopry.

[the remark this answers is not in the document]
Comment (dopry) [reply oauth2_provider/contrib/ninja/security.py:33]: Thanks for taking the time to look into this. I'll try to find the time to consider it in more depth in the next few days.

[the remark this answers is not in the document]
Comment (dopry) [reply oauth2_provider/contrib/ninja/security.py:33]: alright, gonna trust you on this... I'll be tagging you if there are any ninja related issues.

**Comment by dopry** [comment]
Thanks again for this, @brianhelba. To get it merged without waiting, I rebased your branch onto current `master` and opened it as #1727 (your commit authorship is preserved).

On top of the rebase, #1727 resolves the outstanding `__call__` concern: `verify_request` parses the full request URI, so a malformed query string raised an unhandled `ValueError` (HTTP 500) even for header tokens. It now raises `SuspiciousOperation` (HTTP 400) like the DRF integration, with a `test_invalid_hex_string_in_query` regression test. Test deps also moved to `pyproject.toml` + `uv.lock` to match current `master`.

Tracking the merge in #1727 — will close this once it lands.

**Comment by dopry** [comment]
https://github.com/django-oauth/django-oauth-toolkit/pull/1727 landed
